HIP-3.md

```diff
@@ -0,0 +1,189 @@
+- It’s not compatible with situations that involve an extended period between off-chain gas payment estimation and message dispatching, for example messages dispatched by governance, multisig, or timelock contracts. Because top-up payments can be made by any address, these types of messages would likely require a separate payment, which is a poor user experience.
+- There are no refunds for overpayments. Because all data and calculation happens off chain, when a payment is made on the origin chain there is no way to determine if a message has been overpaid for. After a message has been relayed, the relayer could potentially refund overpayments, but this incurs a transaction fee. It’s also not very clear which address should receive the refund.
+- It’s hard for applications to pay for message fees on their user’s behalf. Some contracts are willing to keep a native token balance that they’ll use to pay for message fees. However if the fee amount is calculated off-chain and passed into the contract by a user, there’s nothing stopping a user from specifying a very high fee and depleting the contract’s balance. This approach of contracts paying for fees on a user’s behalf is one way of supporting callback messages.
+- Developers are required to use the Hyperlane SDK to estimate correct payments.
+- Generally, it’s a rough user experience. Application developers have an easier time reasoning about and dealing with on-chain data, and want to have certainty that messages will be processed. Requiring off-chain fee calculation limits the composability of Hyperlane. Sending a message should “just work.”
+
+### **Tech Spec**
+
+Paying for fees can be done by paying an `InterchainGasPaymaster` contract relating to an off-chain relayer that will enter a social contract to process a message.
+
+At a high level, the `InterchainGasPaymaster` contract should accept the message ID of a Hyperlane message that has been dispatched, information about how much gas the relayer should use when processing the message, and payment in the form of native tokens.
+
+Below is the proposed interface for an `InterchainGasPaymaster` and descriptions of what the functions would do for a paymaster that quotes payments on chain. Note that an `InterchainGasPaymaster` with the same function interface could be implemented that doesn’t involve any on-chain fee quoting and still requires off-chain fee calculation.
+
+```solidity
+interface IInterchainGasPaymaster {
+    enum IsmType {
+        Default,
+        NonDefault
+    }
+		
+	/**
+     * @notice Emitted when a gas payment is made.
+     * @param messageId The ID of the message the payment is for.
+     * @param amount The amount of native tokens paid.
+     * @param gas The amount of destination gas paid for.
+     * @param ismType What type of ISM the message expects to be used.
+     */
+    event GasPayment(
+        bytes32 indexed messageId,
+        uint256 amount,
+        uint256 gas,
+        IsmType ismType
```

Comment:
Why is this parameter neccessary?

Reply:
as it's written, there's slightly different relayer behavior based off the ism type